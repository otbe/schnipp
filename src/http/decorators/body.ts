import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Body = () =>
  createMethodParameterDecorator(e => JSON.parse(e.body || '{}'));
