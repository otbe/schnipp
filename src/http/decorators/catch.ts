import { inject } from 'simple-ts-di';

const CATCH = Symbol('catch');

export const Catch = (...exceptions: any[]): ClassDecorator => (
  target: any
) => {
  inject()(target);
  Reflect.defineMetadata(CATCH, exceptions, target.constructor);
};

export const getCaughtExceptions = (catcher: any) =>
  Reflect.getMetadata(CATCH, catcher.constructor) || [];
