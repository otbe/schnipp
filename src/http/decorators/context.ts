import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Context = () =>
  createMethodParameterDecorator((e, m, context) => context);
