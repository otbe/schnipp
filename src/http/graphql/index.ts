export { GraphQLController } from './GraphQLController';
export { Query } from './decorators/query';
export { Mutation } from './decorators/mutation';
export { ResolveField } from './decorators/resolveField';
export { ResolveType } from './decorators/resolveType';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';
