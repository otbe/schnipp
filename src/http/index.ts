import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
export { MetaData } from './utils/MetaData';
export {
  DecoratedExceptionFilter,
  ExceptionFilter
} from './utils/ExceptionFilter';
export { Guard } from './utils/Guard';
export { DefaultExecutionContext } from './utils/ExecutionContext';
export { Catch } from './decorators/catch';
