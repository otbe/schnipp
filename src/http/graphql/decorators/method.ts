import { getResolverMetaData } from './resolver';
import { Newable } from 'simple-ts-di';
import { DecoratedExceptionFilter, Guard, MetaData } from '../..';
import { DocumentNode } from 'graphql';

export const METHOD = Symbol('method');

export type Fields = {
  [fieldName: string]: ResolverMethodMeta<any>;
};

export type ResolverMeta = {
  resolverMap: { [typeName: string]: Fields };
};

export type ResolverMethodMeta<T> = {
  clazz?: Newable<T>;
  methodName: string;
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export const getMethodMetaData = (target: any, methodName: string) => {
  let data: ResolverMethodMeta<any> | null = Reflect.getMetadata(
    METHOD,
    target,
    methodName
  );

  if (data == null) {
    data = {
      methodName,
      filters: [],
      guards: [],
      metaData: {}
    };

    Reflect.defineMetadata(METHOD, data, target, methodName);
  }

  return data;
};

export const createMethodDecorator = (typeName: string) => (
  fieldName: string
): MethodDecorator => <T>(
  resolver: any,
  propertyKey: string | symbol,
  descriptor: TypedPropertyDescriptor<T>
) => {
  const resolverMeta = getResolverMetaData(resolver.constructor);
  resolverMeta.resolverMap[typeName] = resolverMeta.resolverMap[typeName] || {};

  const methodMetaData = getMethodMetaData(resolver, propertyKey.toString());

  methodMetaData.clazz = resolver.constructor;
  resolverMeta.resolverMap[typeName][fieldName] = methodMetaData;

  return descriptor;
};
