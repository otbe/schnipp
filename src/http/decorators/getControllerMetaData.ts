import { MetaData, Guard, DecoratedExceptionFilter } from '..';
import { Newable } from 'simple-ts-di';
import { RestController } from '../rest';
import { GraphQLController } from '../graphql';
import { ResolverMeta } from '../graphql/decorators/method';

export const CONTROLLER = Symbol('controller');

export type ControllerBaseData = {
  headers: { [key: string]: string };
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export type GraphQLControllerData = ControllerBaseData & {
  resolvers?: Array<ResolverMeta>;
};

export type RestControllerData = ControllerBaseData & {
  basePath?: string;
};

export type ControllerData = GraphQLControllerData | RestControllerData;

export const getControllerMetaData = (
  controller: typeof RestController | typeof GraphQLController
) => {
  let data: ControllerData | null = Reflect.getMetadata(CONTROLLER, controller);

  if (data == null) {
    data = {
      filters: [],
      guards: [],
      metaData: {},
      headers: {}
    };

    Reflect.defineMetadata(CONTROLLER, data, controller);
  }

  return data;
};
