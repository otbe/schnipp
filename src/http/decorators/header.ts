import { getControllerMetaData } from './getControllerMetaData';
import { getMethodMetaData } from '../factories/createHttpMethodDecorator';

export const Header = (name: string, value: string) => (
  ...args: Array<any>
) => {
  const target = args[0];

  switch (args.length) {
    // class
    case 1:
      const controllerMetaData = getControllerMetaData(target);
      controllerMetaData.headers[name] = value;
      return;
    // method
    case 2:
    case 3:
      const methodMetaData = getMethodMetaData(args[1], target);
      methodMetaData.headers[name] = value;
      return args[2];
    default:
      throw 'not supported target for @Header';
  }
};
