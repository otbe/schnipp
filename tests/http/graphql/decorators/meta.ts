import 'reflect-metadata';
import { Meta, GraphQLController } from '../../../../src/http/graphql';
import { getControllerMetaData } from '../../../../src/http/decorators/getControllerMetaData';
import { getMethodMetaData } from '../../../../src/http/graphql/decorators/method';

describe('@Meta', () => {
  it('should add metadata for classes on GraphQLControllers', () => {
    @Meta('foo', 'bar')
    class Test extends GraphQLController {
      @Meta('bar', 'baz')
      test() {}

      getResolvers() {
        return [];
      }
    }
    const f = new Test();
    expect(getMethodMetaData(f, 'test')).toMatchSnapshot();
    expect(getControllerMetaData(Test)).toMatchSnapshot();
  });
});
