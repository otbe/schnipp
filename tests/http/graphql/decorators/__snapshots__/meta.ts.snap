// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Meta should add metadata for classes on GraphQLControllers 1`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "metaData": Object {
    "bar": "baz",
  },
  "methodName": "test",
}
`;

exports[`@Meta should add metadata for classes on GraphQLControllers 2`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {
    "foo": "bar",
  },
}
`;
